File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/model/PolarFirstTimeUseConfig.swift
import Foundation
import SwiftProtobuf

public struct PolarFirstTimeUseConfig {

    public enum Gender {
        case male
        case female
    }

    public enum TypicalDay: Int, CaseIterable {
        case mostlySitting = 1
        case mostlyStanding = 2
        case mostlyMoving = 3

        public var description: String {
            switch self {
            case .mostlySitting:
                return "Mostly Sitting"
            case .mostlyStanding:
                return "Mostly Standing"
            case .mostlyMoving:
                return "Mostly Moving"
            }
        }
    }

    public enum TrainingBackground: Int {
            case occasional = 10
            case regular = 20
            case frequent = 30
            case heavy = 40
            case semiPro = 50
            case pro = 60
        }

    public let gender: Gender
    public let birthDate: Date
    public let height: Float
    public let weight: Float
    public let maxHeartRate: Int
    public let vo2Max: Int
    public let restingHeartRate: Int
    public let trainingBackground: TrainingBackground
    public let deviceTime: String
    public let typicalDay: TypicalDay
    public let sleepGoalMinutes: Int

    static let FTU_CONFIG_FILEPATH = "/U/0/S/PHYSDATA.BPB"

    public init(
        gender: Gender,
        birthDate: Date,
        height: Float,
        weight: Float,
        maxHeartRate: Int,
        vo2Max: Int,
        restingHeartRate: Int,
        trainingBackground: TrainingBackground,
        deviceTime: String,
        typicalDay: TypicalDay,
        sleepGoalMinutes: Int
    ) {
        assert(height >= 90.0 && height <= 240.0, "Height must be between 90 and 240 cm")
        assert(weight >= 15.0 && weight <= 300.0, "Weight must be between 15 and 300 kg")
        assert(maxHeartRate >= 100 && maxHeartRate <= 240, "Max heart rate must be between 100 and 240 bpm")
        assert(restingHeartRate >= 20 && restingHeartRate <= 120, "Resting heart rate must be between 20 and 120 bpm")
        assert(trainingBackground.rawValue >= 10 && trainingBackground.rawValue <= 60, "Training background out of range")
        assert(vo2Max >= 10 && vo2Max <= 95, "VO2 max must be between 10 and 95")
        assert(sleepGoalMinutes >= 300 && sleepGoalMinutes <= 660, "Sleep goal must be between 300 and 660 minutes")

        self.gender = gender
        self.birthDate = birthDate
        self.height = height
        self.weight = weight
        self.maxHeartRate = maxHeartRate
        self.vo2Max = vo2Max
        self.restingHeartRate = restingHeartRate
        self.trainingBackground = trainingBackground
        self.deviceTime = deviceTime
        self.typicalDay = typicalDay
        self.sleepGoalMinutes = sleepGoalMinutes
    }

    func toProto() -> Data_PbUserPhysData? {
        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime]
        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)

        guard let deviceTimeDate = dateFormatter.date(from: deviceTime) else {
            BleLogger.error("Failed to parse deviceTime from: \(deviceTime)")
            return nil
        }

        let calendar = Calendar(identifier: .gregorian)
        let utcTimeZone = TimeZone(secondsFromGMT: 0)!
        let components = calendar.dateComponents(in: utcTimeZone, from: deviceTimeDate)

        let lastModified = PbSystemDateTime.with {
            $0.date = PbDate.with {
                $0.year = UInt32(components.year!)
                $0.month = UInt32(components.month!)
                $0.day = UInt32(components.day!)
            }
            $0.time = PbTime.with {
                $0.hour = UInt32(components.hour!)
                $0.minute = UInt32(components.minute!)
                $0.seconds = UInt32(components.second!)
            }
            $0.trusted = true
        }

        let birthday = Data_PbUserBirthday.with {
            $0.value = PbDate.with {
                $0.year = UInt32(Calendar.current.component(.year, from: birthDate))
                $0.month = UInt32(Calendar.current.component(.month, from: birthDate))
                $0.day = UInt32(Calendar.current.component(.day, from: birthDate))
            }
            $0.lastModified = lastModified
        }

        let genderPb = Data_PbUserGender.with {
            $0.value = (gender == .male) ? .male : .female
            $0.lastModified = lastModified
        }

        let weightPb = Data_PbUserWeight.with {
            $0.value = weight
            $0.lastModified = lastModified
        }

        let heightPb = Data_PbUserHeight.with {
            $0.value = height
            $0.lastModified = lastModified
        }

        let maxHeartRatePb = Data_PbUserHrAttribute.with {
            $0.value = UInt32(maxHeartRate)
            $0.lastModified = lastModified
        }

        let restingHeartRatePb = Data_PbUserHrAttribute.with {
            $0.value = UInt32(restingHeartRate)
            $0.lastModified = lastModified
        }

        let trainingBackgroundPb = Data_PbUserTrainingBackground.with {
            $0.value = Data_PbUserTrainingBackground.TrainingBackground(rawValue: trainingBackground.rawValue)!
            $0.lastModified = lastModified
        }

        let vo2MaxPb = Data_PbUserVo2Max.with {
            $0.value = UInt32(vo2Max)
            $0.lastModified = lastModified
        }

        let typicalDayPb = Data_PbUserTypicalDay.with {
            $0.value = Data_PbUserTypicalDay.TypicalDay(rawValue: typicalDay.rawValue)!
            $0.lastModified = lastModified
        }

        let sleepGoalPb = Data_PbSleepGoal.with {
            $0.sleepGoalMinutes = UInt32(sleepGoalMinutes)
            $0.lastModified = lastModified
        }

        return Data_PbUserPhysData.with {
            $0.birthday = birthday
            $0.gender = genderPb
            $0.weight = weightPb
            $0.height = heightPb
            $0.maximumHeartrate = maxHeartRatePb
            $0.restingHeartrate = restingHeartRatePb
            $0.trainingBackground = trainingBackgroundPb
            $0.vo2Max = vo2MaxPb
            $0.typicalDay = typicalDayPb
            $0.sleepGoal = sleepGoalPb
            $0.lastModified = lastModified
        }
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/model/PolarFirstTimeUseConfig.kt
package com.polar.sdk.api.model

import java.util.Date
import java.util.Calendar

import fi.polar.remote.representation.protobuf.Types
import java.time.format.DateTimeFormatter
import fi.polar.remote.representation.protobuf.PhysData
import fi.polar.remote.representation.protobuf.PhysData.PbUserTypicalDay.TypicalDay
import java.time.ZoneOffset
import java.time.ZonedDateTime


data class PolarFirstTimeUseConfig(
    val gender: Gender, // MALE or FEMALE
    val birthDate: Date, // String
    val height: Float, // cm, valid range [90-240]
    val weight: Float, // kg, valid range [15-300]
    val maxHeartRate: Int, // bpm, valid range [100-240]
    val vo2Max: Int, // valid Range: [10-95] ml/kg/min]
    val restingHeartRate: Int, // bpm, valid range [20-120]
    val trainingBackground: Int, // valid values [10, 20, 30, 40, 50, 60]
    val deviceTime: String, // ISO 8601 DateTime String
    val typicalDay: TypicalDay, // One of [TypicalDay] values
    val sleepGoalMinutes: Int // Minutes, valid range [300-660]
) {
    enum class Gender {
        MALE, FEMALE;
    }

    enum class TypicalDay(val index: Int) {
        MOSTLY_SITTING(1),
        MOSTLY_STANDING(2),
        MOSTLY_MOVING(3);
    }
    init {
        require(height in HEIGHT_MIN..HEIGHT_MAX) { "Height must be between $HEIGHT_MIN and $HEIGHT_MAX cm" }
        require(weight in WEIGHT_MIN..WEIGHT_MAX) { "Weight must be between $WEIGHT_MIN and $WEIGHT_MAX kg" }
        require(maxHeartRate in MAX_HEART_RATE_MIN..MAX_HEART_RATE_MAX) { "Max heart rate must be between $MAX_HEART_RATE_MIN and $MAX_HEART_RATE_MAX bpm" }
        require(restingHeartRate in RESTING_HEART_RATE_MIN..RESTING_HEART_RATE_MAX) { "Resting heart rate must be between $RESTING_HEART_RATE_MIN and $RESTING_HEART_RATE_MAX bpm" }
        require(vo2Max in VO2_MAX_MIN..VO2_MAX_MAX) { "VO2 max must be between $VO2_MAX_MIN and $VO2_MAX_MAX" }
        require(trainingBackground in TRAINING_BACKGROUND_VALUES) { "Training background must be one of the following values: ${TRAINING_BACKGROUND_VALUES.joinToString()}" }
        require(sleepGoalMinutes in SLEEP_GOAL_RANGE_MINUTES) { "Sleep goal must be between ${SLEEP_GOAL_RANGE_MINUTES.first} and ${SLEEP_GOAL_RANGE_MINUTES.last} minutes" }
    }

    companion object {
        const val HEIGHT_MIN = 90.0f
        const val HEIGHT_MAX = 240.0f
        const val WEIGHT_MIN = 15.0f
        const val WEIGHT_MAX = 300.0f
        const val MAX_HEART_RATE_MIN = 100
        const val MAX_HEART_RATE_MAX = 240
        const val RESTING_HEART_RATE_MIN = 20
        const val RESTING_HEART_RATE_MAX = 120
        const val VO2_MAX_MIN = 10
        const val VO2_MAX_MAX = 95
        val TRAINING_BACKGROUND_VALUES = listOf(10, 20, 30, 40, 50, 60)
        val SLEEP_GOAL_RANGE_MINUTES = 300..660
        const val FTU_CONFIG_FILENAME = "/U/0/S/PHYSDATA.BPB"
    }
}

    fun PolarFirstTimeUseConfig.toProto(): PhysData.PbUserPhysData {
    val deviceTimeUtc = ZonedDateTime.parse(deviceTime, DateTimeFormatter.ISO_DATE_TIME)
            .withZoneSameInstant(ZoneOffset.UTC)
    val lastModified = Types.PbSystemDateTime.newBuilder()
            .setDate(Types.PbDate.newBuilder()
                    .setYear(deviceTimeUtc.year)
                    .setMonth(deviceTimeUtc.monthValue)
                    .setDay(deviceTimeUtc.dayOfMonth)
                    .build())
            .setTime(Types.PbTime.newBuilder()
                    .setHour(deviceTimeUtc.hour)
                    .setMinute(deviceTimeUtc.minute)
                    .setSeconds(deviceTimeUtc.second)
                    .build())
            .setTrusted(true)
            .build()

        val birthdayParsed = Calendar.getInstance().apply { time = birthDate }
        val birthday = PhysData.PbUserBirthday.newBuilder().apply {
            setValue(Types.PbDate.newBuilder()
                .setYear(birthdayParsed.get(Calendar.YEAR))
                .setMonth(birthdayParsed.get(Calendar.MONTH) + 1)
                .setDay(birthdayParsed.get(Calendar.DAY_OF_MONTH))
                .build())
            setLastModified(lastModified)
        }.build()

        val gender = PhysData.PbUserGender.newBuilder().apply {
            setValue(when (gender) {
                PolarFirstTimeUseConfig.Gender.MALE -> PhysData.PbUserGender.Gender.MALE
                PolarFirstTimeUseConfig.Gender.FEMALE -> PhysData.PbUserGender.Gender.FEMALE
            })
            setLastModified(lastModified)
        }.build()

        val weightBuilder = PhysData.PbUserWeight.newBuilder().apply {
            setValue(weight)
            setLastModified(lastModified)
        }.build()

        val heightBuilder = PhysData.PbUserHeight.newBuilder().apply {
            setValue(height)
            setLastModified(lastModified)
        }.build()

        val maxHeartRateBuilder = PhysData.PbUserHrAttribute.newBuilder().apply {
            setValue(maxHeartRate)
            setLastModified(lastModified)
        }.build()

        val restingHeartRateBuilder = PhysData.PbUserHrAttribute.newBuilder().apply {
            setValue(restingHeartRate)
            setLastModified(lastModified)
        }.build()

        val trainingBackgroundBuilder = PhysData.PbUserTrainingBackground.newBuilder().apply {
            setValue(PhysData.PbUserTrainingBackground.TrainingBackground.forNumber(trainingBackground))
            setLastModified(lastModified)
        }.build()

        val vo2MaxBuilder = PhysData.PbUserVo2Max.newBuilder().apply {
            setValue(vo2Max) // Range: [10-95] ml/kg/min
            setLastModified(lastModified)
        }.build()

        val typicalDay = PhysData.PbUserTypicalDay.newBuilder()
                .setValue(TypicalDay.forNumber(typicalDay.index))
                .setLastModified(lastModified)
                .build()

        val sleepGoal = PhysData.PbSleepGoal.newBuilder()
                .setSleepGoalMinutes(sleepGoalMinutes)
                .setLastModified(lastModified)
                .build()

        return PhysData.PbUserPhysData.newBuilder().apply {
            setBirthday(birthday)
            setGender(gender)
            setWeight(weightBuilder)
            setHeight(heightBuilder)
            setMaximumHeartrate(maxHeartRateBuilder)
            setRestingHeartrate(restingHeartRateBuilder)
            setTrainingBackground(trainingBackgroundBuilder)
            setVo2Max(vo2MaxBuilder)
            setTypicalDay(typicalDay)
            setSleepGoal(sleepGoal)
            setLastModified(lastModified)
        }.build()
    }


